src/Lecture2.hs

```diff
@@ -70,7 +79,14 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt:: Int -> [a] -> (Maybe a, [a])
+removeAt idx list = removeAt_ 0 list []
+  where
+    removeAt_ _ [] list_ = (Nothing, list_)
+    removeAt_ idx_ (x : xs) list_
+        | idx == idx_   = (Just x, list_ ++ xs)
+        | otherwise     = removeAt_ (idx_ + 1) xs (list_ ++ [x])
```

Comment:
That's a correct implementation 👍🏻 

However, it is more efficient to construct lists using the cons operator (`:`) directly without introducing intermediate values or adding lists with `(++)` whenever possible. Adding elements to the end of the list is rather slow as well 🐌

Instead of adding elements to the end, you can get the result recursively and add to it. Something like this:

```haskell
removeAt n (x : xs) =
    let (value, rest) = removeAt (n - 1) xs
    in (value, x : rest)
```

I also noticed that your current implementation hangs when its given a negative number and an infinite list ♾️ Probably you can handle negative numbers separately using a special check.

Alternatively, you can use a standard function [splitAt](https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:splitAt) instead of writing manual recursion 🔃